### PR TITLE
Gseal 105 display most popular articles

### PR DESCRIPTION
**Pull request by gpseal**
Attempting to display most clicked stories in the article.summary route, allowing most popular articles to be displayed on the home page of the site.

When a user visits a page, a record is added to the database in a table called 'Article_Visits' which stores number of clicks, plus the date clicked.  Articles fetched with the article.summary route will now be displayed ranked by clicks within this table (or by date if there are no clicks).  Only clicks from the past three days will be used, to keep things relevant.

I will adjust to prevent clicks being added with page refreshes (next sprint).

### CODE REVIEW

File: api/route/Article/article.summary.ts [hunk 1]
```diff
@@ -4,16 +4,22 @@ import { getClient } from '../../database';
...
+const ARTICLES_PER_CATEGORY = 3;
+
+// query shows three items from each category ranking from most clicks to least, 
+// if no clicks registered for articles, orders by publication date
 const groupCategoriesInThrees = `
-  select id, name, author, publication_date, category, source_url, cover_url
-  from (
-    select ar.id, a.name, ar.author, a.body, ar.category, ar.source_url, ar.cover_url, ar.retrieved_date, ar.publication_date, 
-      row_number() over (PARTITION BY ar.category order by ar.publication_date) index_in_category
+  select * from (
```
Comment (AardhynLavender) [inline api/route/Article/article.summary.ts:12]: can we swap `*` out with just the fields we need. Even if we need all of them its still faster to list them explicitly over using `*`.

File: api/route/Article/article.record_clicks.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      // records clicks using stored procedure 'record_visit'
+      await getClient().query(`call record_visit($1, $2)`, [id, today]); // sends queries
+      return true;
+    } catch (error) {
+      console.error(error);
+      return false;
+    }
+  };
+
+  addRecord(id).then(result => {
```
Comment (AardhynLavender) [inline api/route/Article/article.record_clicks.ts:23]: Can we swap out the these method calls for the `try` `catch` syntax.

**Review by AardhynLavender — CHANGES_REQUESTED** [review]
Everything looks good. Just a couple of minor things, really.

File: database/src/init.sql [hunk 3]
```diff
@@ -141,5 +150,18 @@ values ('Michael Knight'),
...
-        
\ No newline at end of file
+create or replace procedure record_visit(
+   sent_id integer,
+   today date
+)
+language plpgsql    
+as $$
+begin
+if exists (select * from Article_Visits where sent_id = id and click_date = today) then
```
[the remark this answers is not in the document]
Comment (ArtemKech) [reply database/src/init.sql:160]: Agreed with @AardhynLavender. Good thing to change

File: api/route/Article/article.record_clicks.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    if (result) {
+      console.log('Visit recorded');
```
Comment (ArtemKech) [inline api/route/Article/article.record_clicks.ts:25]: Maybe get rid of this line if it was made for debugging purposes

File: client/src/pages/ArticlePage.tsx [hunk 5]
```diff
@@ -26,6 +27,10 @@ const ArticlePage = () => {
         getData("/category.list", setCategories);
         window.scrollTo(0, 0);
         },[])
+    
+    useEffect(() => {
```
Comment (ArtemKech) [inline client/src/pages/ArticlePage.tsx:31]: There are 3 useEffects there. Can you call `postData` in one of them to reduce redundancy? Or there is a reason for them to be in different ones?

File: database/src/init.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (gpseal) [reply database/src/init.sql:160]: good point

File: api/route/Article/articleRouter.ts [hunk 6]
```diff
@@ -45,6 +48,14 @@ const articleRouter = createRouter('article', {
     method: 'get',
     handler: getAll,
   },
+  record_visit: {
```
[the remark this answers is not in the document]
Comment (gpseal) [reply api/route/Article/articleRouter.ts:51]: I was going to use a record in local storage to try to stop the refreshing page issue (though determined users can still clear browsing data to get around this)

File: client/src/pages/ArticlePage.tsx (hunk 5 above)
Comment (gpseal) [reply client/src/pages/ArticlePage.tsx:31]: have merged two of them into one

**Review by AardhynLavender — APPROVED** [review]

**Review by ArtemKech — APPROVED** [review]
Really like this feature. Great work! Makes our app look more professional 😎
Thanks for fixing requested files

**Pull request merged**